lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -55,77 +55,74 @@ void ExecutionEngine::setBackend(Backend *backend) {
 
 ExecutionEngine::~ExecutionEngine() = default;
 
-void ExecutionEngine::run(llvm::ArrayRef<Variable *> vars,
-                          llvm::ArrayRef<Tensor *> inputs) {
-  assert(function_ && "No function has been compiled");
+void glow::updateVariables(llvm::ArrayRef<Variable *> vars,
+                           llvm::ArrayRef<Tensor *> inputs) {
   assert(inputs.size() == vars.size() &&
          "The number of inputs does not match the number of variables");
 
   // Update the input variables.
   for (int i = 0, e = vars.size(); i < e; i++) {
+    assert(vars[i] && "Invalid value");
     assert(vars[i]->getVisibilityKind() == VisibilityKind::Public &&
            "Trying to update a private variable");
-    loadValueFromTensor(vars[i], inputs[i]);
+    auto &t = vars[i]->getPayload();
+    auto dim = inputs[i]->dims();
+    (void)dim;
+    assert(t.dims() == dim &&
+           t.getElementType() == inputs[i]->getElementType() &&
+           "Mismatch on Variable and Tensor types.");
+    t.assign(inputs[i]);
   }
+}
 
+void ExecutionEngine::run() {
+  assert(function_ && "No function has been compiled");
   function_->execute();
 }
 
-void ExecutionEngine::runBatch(size_t iterations,
-                               llvm::ArrayRef<Variable *> vars,
-                               llvm::ArrayRef<Tensor *> inputs) {
-  static size_t trainCounter = 0;
-
-  assert(function_ && "No function has been compiled");
+/// Update the content of the tensors \p vars with some slices that are from \p
+/// inputs. The data starts at slice \p sampleIdx and wraps around until the
+/// data in \p v is filled. All dimensions, except for the first (batch)
+/// dimension must be identical.
+void glow::updateVariablesFromBatch(llvm::ArrayRef<Variable *> vars,
+                                    llvm::ArrayRef<Tensor *> inputs,
+                                    size_t sampleIdx) {
   assert(!inputs.empty() && "No inputs");
   assert(inputs.size() == vars.size() &&
          "The number of inputs does not match the number of variables");
 
+  // Update the input variables.
+  for (int i = 0, e = vars.size(); i < e; i++) {
+    assert(vars[i] && "Invalid value");
+    auto &t = vars[i]->getPayload();
+
+    auto dim = inputs[i]->dims();
+    assert(t.dims().drop_front() == dim.drop_front() && "Invalid slice size");
+    // Extract the n'th slice, that must be a tensor.
+    size_t slc = sampleIdx % dim[0];
+    t.copyConsecutiveSlices(inputs[i], slc);
+  }
+}
+
+void glow::runBatch(ExecutionEngine &EE, size_t iterations,
+                    llvm::ArrayRef<Variable *> vars,
+                    llvm::ArrayRef<Tensor *> inputs) {
+  static size_t trainCounter = 0;
```

Comment:
why `trainCounter` needs to be static? Can you please explain.

Reply:
I added a big comment that explains why this is a really bad idea to have this global variable. I think that I have a quick fix to get rid of the global variable. Basically, now that this is only a helper function we can change the call sites and make it a parameter.